binary_search_tree/tree.py

```diff
@@ -7,52 +7,134 @@ def __init__(self, key, val = None):
+        
+        return current_root 
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(log n)
     def add(self, key, value = None):
-        pass
+        # if there is no root...
+        if not self.root:
+            self.root = TreeNode(key, value) 
+            return 
+
+        new_node = TreeNode(key, value)
+        self.add_helper(self.root, new_node)
 
-    # Time Complexity: 
-    # Space Complexity: 
+    def find_helper(self, current_root, target_key):
+        # Base cases: root is null or key is present at root
+        if current_root is None or current_root.key == target_key:
+            return current_root 
+    
+        # Search left 
+        if current_root.key >= target_key:
+          return self.find_helper(current_root.left, target_key)  
+        
+        # Search right 
+        return self.find_helper(current_root.right, target_key)
+      
+    # Time Complexity: O(log n)
+    # Space Complexity: O(log n)
     def find(self, key):
-        pass
+        target_node = self.find_helper(self.root, key)
+
+        if target_node is None:
+            return None 
 
-    # Time Complexity: 
-    # Space Complexity: 
+        return target_node.value 
+
+    # helpers 
+    def inorder_helper(self, current_node, items):
+        if current_node is not None:
+            self.inorder_helper(current_node.left, items) 
+            items.append({ "key" : current_node.key, "value" : current_node.value })
+            self.inorder_helper(current_node.right, items) 
+
+    def preorder_helper(self, current_node, items):
+        if current_node is not None:
+            items.append({ "key" : current_node.key, "value" : current_node.value })
+            self.preorder_helper(current_node.left, items) 
+            self.preorder_helper(current_node.right, items) 
+    
+    def postorder_helper(self, current_node, items):
+        if current_node is not None:
+            self.postorder_helper(current_node.left, items) 
+            self.postorder_helper(current_node.right, items) 
+            items.append({ "key" : current_node.key, "value" : current_node.value })
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def inorder(self):
-        pass
+        items = []
+        self.inorder_helper(self.root, items)
+        return items 
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)   
     def preorder(self):
-        pass
+        items = []
+        self.preorder_helper(self.root, items)
+        return items 
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)  
     def postorder(self):
-        pass
+        items = []
+        self.postorder_helper(self.root, items)
+        return items
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(n) or O(log n) ???
     def height(self):
```

Comment:
👍  Correct the space complexity due to the call stack is O(n) if the tree is unbalanced and O(log n) if the tree is balanced.